test/fixtures/es-module-loaders/mock-loader.mjs

```diff
@@ -0,0 +1,105 @@
+import {receiveMessageOnPort} from 'worker_threads';
+const mockedModuleExports = new Map();
+let currentMockVersion = 0;
+
+/**
+ * FIXME: this is a hack to workaround loaders being
+ * single threaded for now
+ */
+function doDrainPort() {
+  let msg;
+  while (msg = receiveMessageOnPort(preloadPort)) {
+    onPreloadPortMessage(msg.message);
+  }
+}
+function onPreloadPortMessage({
+  mockVersion, resolved, exports
+}) {
+  currentMockVersion = mockVersion;
+  mockedModuleExports.set(resolved, exports);
+}
+let preloadPort;
+export function getGlobalPreloadCode({port}) {
+  preloadPort = port;
+  port.on('message', onPreloadPortMessage);
+  return `
+    let mockedModules = new Map();
+    let mockVersion = 0;
+    globalThis.mock = (resolved, replacementProperties) => {
+      let exports = Object.keys(replacementProperties);
+      let shadow = Object.create(null);
+      let namespace = Object.create(null);
+      for (const name of exports) {
+        Object.defineProperty(namespace, name, {
+          get() {
+            return shadow[name];
+          },
+          set(v) {
+            shadow[name] = v;
+          }
+        });
+      }
+      mockedModules.set(resolved, replacementProperties);
+      mockVersion++;
+      port.postMessage({mockVersion, resolved, exports });
+      return namespace;
+    }
+    setImportMetaCallback((meta, context, parent) => {
+      if (context.url.startsWith('mock:')) {
+        let [proto, version, encodedTargetURL] = context.url.split(':');
+        let decodedTargetURL = decodeURIComponent(encodedTargetURL);
+        if (mockedModules.has(decodedTargetURL)) {
+          meta.mock = mockedModules.get(decodedTargetURL);
+          return;
+        }
+      }
+      parent(meta, context);
+    });
+    globalThis.mock.map = mockedModules;`;
```

Comment:
Might this be better as a fixture?

Reply:
I'm fine making a fixture if desired, but this method returns a string and creating a fixture that isn't meant to be reused since it relies on the behavior in this `mock-loader.mjs` file seems a bit odd to me. Is there some gain to be had? we could always wrap this in a IIFE and template literal if you want syntax highlighting I guess.